### PR TITLE
Update panopticnet and fpn for future model export support

### LINKED ISSUE #338: All layers of panopticnets should be named
**Is your feature request related to a problem? Please describe.**
All of the layers of panopticnets should be named. If names aren't defined, then layers are given numeric id's from tensorflow (e.g. tensor_product_2, etc) and the id is based on the number of layers of that type that are already in existence. These names get saved when saving models, which makes transfer learning more difficult. For example, without naming each layer, we can't load weights by name.

**Describe the solution you'd like**
Explicit names for every layer in panoptic nets.

**Additional context**
Image normalization and location layers have to be removed to create models that can be exported with tflite. Its much easier to just create new models without these layers and load weights as opposed to trying to perform surgery on an existing model.

### PR DESCRIPTION
**Pull request by msschwartz21**
## What
- Removed TensorProduct for Conv's in PanopticNet
- Added names to all the layers of PanopticNet (Closes #338)
- Removed semantic_prediction which wasn't being used in the file
- Added an option to use depthwise conv's in feature pyramid construction as opposed to regular convs

## Why
* TensorProduct not supported for TfLite
* Names required for saved model export

### CODE REVIEW

**Comment by willgraf** [comment]
Are `Conv` layers the right substitute for `TensorProduct`? I think I remember that they are more like the `Dense` layers.

File: deepcell/model_zoo/fpn.py [hunk 1]
```diff
@@ -291,20 +306,26 @@ def semantic_upsample(x, n_upsample, n_filters=64, ndim=2, target=None):
...
+            x = upsampling(size=size,
+                           name='upsampling_{}_semantic_upsample_{}'.format(i, semantic_id),
+                           interpolation='bilinear')(x)
 
     if n_upsample == 0:
-        x = conv(n_filters, 3, strides=1,
-                 padding='same', data_format='channels_last')(x)
+        x = conv(n_filters, conv_kernel, strides=1,
+                 padding='same', data_format='channels_last',
+                 name='conv_final_semantic_upsample_{}'.format(semantic_id))(x)
```
Comment (msschwartz21) [inline deepcell/model_zoo/fpn.py:328]: @vanvalen `semantic_id` is not defined in this function.

**Comment by vanvalen** [comment]
Or you could add semantic_id=0 - the key changes for semantic heads are in the panopticnet.py file.

**Comment by vanvalen** [comment]
Looks good!

File: deepcell/model_zoo/panopticnet.py [hunk 2]
```diff
@@ -236,6 +168,10 @@ def __create_semantic_head(pyramid_dict,
     Returns:
         keras.layers.Layer: The semantic segmentation head
     """
+
+    conv = Conv2D if ndim == 2 else Conv3D
+    conv_kernel = (1, 1) if ndim == 2 else (1, 1, 1)
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:173]: Could make this cooler like `conv_kernel = (1,) * ndim`
---
assuming `ndim` is either 2 or 3.

File: deepcell/model_zoo/panopticnet.py [hunk 3]
```diff
@@ -252,21 +188,25 @@ def __create_semantic_head(pyramid_dict,
...
+    semantic_name = pyramid_names[-1]
 
     # Final upsampling
-    min_level = int(re.findall(r'\d+', semantic_names[-1])[0])
+    min_level = int(re.findall(r'\d+', semantic_name[-1])[0])
     n_upsample = min_level
-    x = semantic_upsample(semantic_sum, n_upsample, ndim=ndim)
+    x = semantic_upsample(semantic_feature, n_upsample, ndim=ndim, semantic_id=semantic_id)
 
     # First tensor product
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:199]: These comments don't seem right anymore.

File: deepcell/model_zoo/panopticnet.py [hunk 4]
```diff
@@ -355,35 +300,36 @@ def PanopticNet(backbone,
...
     else:
         concat = norm
 
-    if frames_per_batch > 1:
-        fixed_inputs = TimeDistributed(TensorProduct(required_channels))(concat)
-    else:
-        fixed_inputs = TensorProduct(required_channels)(concat)
+    # Force the channel size for backbone input to be `required_channels`
+    fixed_inputs = conv(required_channels, conv_kernel, strides=1,
+                        padding='same', name='conv_channels')(concat)
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:330]: Is this the same thing? I thought that `TensorProduct` is more like a `Dense` layer.

Comment (willgraf) [reply deepcell/model_zoo/panopticnet.py:330]: I am mistaken.

`TensorProduct` is the mathematical equivalent to a `Conv` with a 1x1 filter, which is why we instantiate `conv_kernel = (1, 1)` (for 2D).  `Dense` layers, on the other hand, flatten the inputs, which is not desired in this case.

File: deepcell/model_zoo/panopticnet.py (hunk 2 above)
Comment (msschwartz21) [reply deepcell/model_zoo/panopticnet.py:173]: The only downside to that change is that we could end up with a 4 or 5d kernel if the user submitted the wrong value for ndim. Perhaps the most appropriate change would be to first check that ndim is 2 or 3?

**Comment by willgraf** [comment]
Any reason we are using two different kwarg names for TFLite export? `lite` vs `lite_fpn`?

**Comment by vanvalen** [comment]
There's no reason why lite_fpn can't be changed to lite - I just changed it.

File: deepcell/model_zoo/fpn.py [hunk 5]
```diff
@@ -79,14 +82,22 @@ def create_pyramid_level(backbone_input,
...
 
+    # Check if inputs to ndim and lite are compatible
     if ndim == 3 and lite:
         raise ValueError('lite == True is not compatible with 3 dimensional networks')
 
+    # Check input to interpolation
+    acceptable_interpolation = {'bilinear', 'nearest'}
+    if interpolation not in acceptable_interpolation:
+        raise ValueError('Interpolation mode not supported. Choose from '
+            '[\'bilinear\', \'nearest\']')
```
Comment (willgraf) [inline deepcell/model_zoo/fpn.py:98]: The strings have to align to pass PEP8, also I suggest using double-quotes inside the string for better readability.

```
        raise ValueError('Interpolation mode not supported. Choose from '
                                   '["bilinear", "nearest"]')
```

File: deepcell/model_zoo/fpn.py [hunk 6]
```diff
@@ -173,6 +187,11 @@ def __create_pyramid_features(backbone_dict,
     if ndim not in acceptable_ndims:
         raise ValueError('Only 2 and 3 dimensional networks are supported')
 
+    acceptable_interpolation = {'bilinear', 'nearest'}
+    if interpolation not in acceptable_interpolation:
+        raise ValueError('Interpolation mode not supported. Choose from '
+            '[\'bilinear\', \'nearest\']')
+
```
Comment (willgraf) [inline deepcell/model_zoo/fpn.py:194]: The strings have to align to pass PEP8, also I suggest using double-quotes inside the string for better readability.

```
        raise ValueError('Interpolation mode not supported. Choose from '
                         '["bilinear", "nearest"]')
```

File: deepcell/model_zoo/panopticnet.py [hunk 7]
```diff
@@ -113,17 +114,26 @@ def semantic_upsample(x, n_upsample, n_filters=64, ndim=2, semantic_id=0):
...
     acceptable_ndims = [2, 3]
     if ndim not in acceptable_ndims:
         raise ValueError('Only 2 and 3 dimensional networks are supported')
 
+    # Check input to interpolation
+    acceptable_interpolation = {'bilinear', 'nearest'}
+    if interpolation not in acceptable_interpolation:
+        raise ValueError('Interpolation mode not supported. Choose from '
+                         '[\'bilinear\', \'nearest\']')
+
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:136]: This one has good spacing, but still suggest the double quotes.

File: deepcell/model_zoo/panopticnet.py [hunk 8]
```diff
@@ -163,18 +174,26 @@ def __create_semantic_head(pyramid_dict,
...
+    # Check input to ndims
     if ndim not in {2, 3}:
         raise(ValueError('ndim must be either 2 or 3. Received ndim = {}'.format(ndim)))
 
+    # Check input to interpolation
+    acceptable_interpolation = {'bilinear', 'nearest'}
+    if interpolation not in acceptable_interpolation:
+        raise ValueError('Interpolation mode not supported. Choose from '
+                         '[\'bilinear\', \'nearest\']')
+
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:196]: One more time.

File: deepcell/model_zoo/panopticnet.py [hunk 9]
```diff
@@ -299,6 +322,12 @@ def PanopticNet(backbone,
             raise ValueError('Mode {} not supported. Please choose from {}.'.format(
                 temporal_mode, str(acceptable_modes)))
 
+    # Check input to interpolation
+    acceptable_interpolation = {'bilinear', 'nearest'}
+    if interpolation not in acceptable_interpolation:
+        raise ValueError('Interpolation mode not supported. Choose from '
+                         '[\'bilinear\', \'nearest\']')
+
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:330]: Last time. We do this check a lot!

File: deepcell/model_zoo/panopticnet.py [hunk 10]
```diff
@@ -114,100 +115,42 @@ def semantic_upsample(x, n_upsample, n_filters=64, ndim=2):
             the 3x3 convolution
         ndim (int): The spatial dimensions of the input data.
             Default is 2, but it also works with 3
+        interpolation (str): Choice of interpolation mode for upsampling
+            layers from ['bilinear', 'nearest'. Defaults to bilinear.
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:119]: Missing closing bracket `]`.

File: deepcell/model_zoo/panopticnet.py [hunk 11]
```diff
@@ -219,23 +162,43 @@ def __create_semantic_head(pyramid_dict,
...
         n_filters (int): Defaults to 64. The number of convolutional filters.
         n_dense (int): Defaults to 128. Number of dense filters.
         semantic_id (int): Defaults to 0.
         ndim (int): Defaults to 2, 3d supported.
         include_top (bool): Defaults to False.
         target_level (int, optional): Defaults to 2. The level we need to reach.
-            Performs 2x upsampling until we're at the target level
+            Performs 2x upsampling until we're at the target level.
+        interpolation (str): Choice of interpolation mode for upsampling
+            layers from ['bilinear', 'nearest'. Defaults to bilinear.
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:180]: Missing closing bracket `]`.

File: deepcell/model_zoo/panopticnet.py [hunk 12]
```diff
@@ -301,19 +271,25 @@ def PanopticNet(backbone,
...
-        norm_method (str): ImageNormalization mode to use. Defaults to 'whole_image'
-        location (bool): Whether to include location data.
+        norm_method (str): ImageNormalization mode to use.
+            Defaults to 'whole_image'.
+        location (bool): Whether to include location data. Defaults to True
         use_imagenet (bool): Whether to load imagenet-based pretrained weights.
+        lite (bool): Whether to use a depthwise conv in the feature pyramid
+            rather than regular conv. Defaults to False.
+        interpolation (str): Choice of interpolation mode for upsampling
+            layers from ['bilinear', 'nearest'. Defaults to bilinear.
```
Comment (willgraf) [inline deepcell/model_zoo/panopticnet.py:292]: Missing closing bracket `]`.
